packages/router/src/router.tsx

```diff
@@ -1,34 +1,129 @@
 // The guts of the router implementation.
 
-import React, { ReactChild, ReactElement } from 'react'
-
-import { useAuth as useAuthHook } from '@redwoodjs/auth'
+import React, { ReactChild } from 'react'
 
 import {
-  Location,
   parseSearch,
   replaceParams,
   matchPath,
-  mapNamedRoutes,
   PageLoader,
   Redirect,
-  LocationContextType,
-  ParamType,
+  useLocation,
+  validatePath,
+  LocationProvider,
 } from './internal'
+import { PrivateContextProvider, usePrivate } from './private-context'
+import {
+  RouterContextProvider,
+  RouterState,
+  useRouterState,
+} from './router-context'
 import { SplashPage } from './splash-page'
+import { flattenAll, isReactElement } from './util'
+
+/**
+ * A more specific interface will be generated by
+ * babel-plugin-redwood-routes-auto-loader when a redwood project is built
+ *
+ * Example:
+ * interface AvailableRoutes {
+ *   home: () => "/"
+ *   test: () => "/test"
+ * }
+ */
+interface AvailableRoutes {
+  [key: string]: (args?: Record<string, string>) => string
+}
+
+const namedRoutes: AvailableRoutes = {}
+
+type PageType =
+  | Spec
+  | React.ComponentType<unknown>
+  | ((props: any) => JSX.Element)
 
 interface RouteProps {
-  page: Spec | React.ComponentType<unknown> | ((props: any) => JSX.Element)
   path?: string
+  page?: PageType
   name?: string
   notfound?: boolean
   redirect?: string
   prerender?: boolean
   whileLoading?: () => ReactChild | null
```

Comment:
We have three kinds of routes:
1. Route: path, page, name
2. Redirect: path, redirect
3. NotFound: notfound, page

There is no prop that is common across all three, so they all have to be optional

Reply:
I was wondering if we could tighten up the "user facing types" and then explicitly ignore the errors internally.

Reply:
The only thing I can think of, is to make it explicit, and have `<Route>`, `<RedirectRoute>` and `<NotFoundRoute>`. Do we want that?

Reply:
Or, what if we had a `MegaInternalRoute` internally with these types, but an external `Route` component with tighter types. I'm trying to figure out a nice way to help the user understand what we want.

Reply:
I managed to make it a bit better, but not as good as we'd want it to be.

```
interface RouteProps {
  path: string
  page: PageType
  name: string
  prerender?: boolean
  whileLoading?: () => React.ReactElement | null
}

interface RedirectRouteProps {
  path: string
  redirect: string
}

interface NotFoundRouteProps {
  notfound: boolean
  page: PageType
}

type InternalRouteProps = Partial<
  RouteProps & RedirectRouteProps & NotFoundRouteProps
>

const Route: React.VFC<RouteProps | RedirectRouteProps | NotFoundRouteProps> = (
  props
) => {
  return <InternalRoute {...props} />
}
```

![image](https://user-images.githubusercontent.com/30793/111628246-4e455700-87f0-11eb-8d29-076e9a81ab21.png)

![image](https://user-images.githubusercontent.com/30793/111628315-60bf9080-87f0-11eb-878c-0bad2e4e3bc0.png)

Those two are good. But it lets this pass, which it would be nice if it wouldn't

![image](https://user-images.githubusercontent.com/30793/111628504-949ab600-87f0-11eb-8609-15aa2ad65b99.png)

How do you want to proceed?